hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -0,0 +1,2259 @@
+package is.hail.expr.ir.streams
+
+import is.hail.annotations.Region
+import is.hail.asm4s._
+import is.hail.expr.ir._
+import is.hail.expr.ir.orderings.StructOrdering
+import is.hail.services.shuffler.{CodeShuffleClient, ShuffleClient, ValueShuffleClient}
+import is.hail.types.physical.stypes.SType
+import is.hail.types.physical.stypes.concrete.{SBinaryPointer, SBinaryPointerSettable, SCanonicalShufflePointerCode, SCanonicalShufflePointerSettable}
+import is.hail.types.physical.stypes.interfaces._
+import is.hail.types.physical.stypes.primitives.{SInt32, SInt32Code}
+import is.hail.types.physical.{PCanonicalStream, PCode, PInterval, PStruct, PType}
+import is.hail.types.virtual.{TInterval, TShuffle, TStream}
+import is.hail.utils._
+
+
+object StreamProducer {
+  def markUnused(producer: StreamProducer, mb: EmitMethodBuilder[_]): Unit = {
+    (producer.LendOfStream.isImplemented, producer.LproduceElementDone.isImplemented) match {
+      case (true, true) =>
+      case (false, false) =>
+
+        EmitCodeBuilder.scopedVoid(mb) { cb =>
+          cb.define(producer.LendOfStream)
+          cb.define(producer.LproduceElementDone)
+          cb._fatal("unreachable")
+        }
+
+      case (eos, ped) => throw new RuntimeException(s"unrealizable value unused asymmetrically: eos=$eos, ped=$ped")
+    }
+    producer.element.pv match {
+      case SStreamCode(_, nested) => markUnused(nested, mb)
+      case _ =>
+    }
+  }
+}
+
+abstract class StreamProducer {
+  /**
+    * Stream length, which is present if it can be computed (somewhat) cheaply without
+    * consuming the stream.
+    *
+    * In order for `length` to be valid, the stream must have been initialized with `initialize`.
+    */
+  val length: Option[Code[Int]]
+
+  /**
+    * Stream producer setup method. If `initialize` is called, then the `close` method
+    * must be called as well to properly handle owned resources like files.
+    *
+    * The stream's element region must be assigned by a consumer before initialize
+    * is called.
+    *
+    * This block cannot jump away, e.g. to `LendOfStream`.
+    *
+    */
+  def initialize(cb: EmitCodeBuilder): Unit
+
+  /**
+    * Stream element region, into which the `element` is emitted. The assignment, clearing,
+    * and freeing of the element region is the responsibility of the stream consumer.
+    */
+  val elementRegion: Settable[Region]
+
+  /**
+    * This boolean parameter indicates whether the producer's elements should be emitted into
+    * separate regions (by clearing when elements leave a consumer's scope). This parameter
+    * propagates bottom-up from producers like [[ReadPartition]] and [[StreamRange]], but
+    * it is the responsibility of consumers to implement the right memory management semantics
+    * based on this flag.
+    */
+  val separateRegions: Boolean
```

Comment:
This is minor, but I'm interpreting this to mean "the elements of this stream cannot safely be stored in memory at once".  Using separate regions is one solution (as is using one region and clearing it).  Rather than naming this after the (a) implementation strategy, I think a descriptive property of the stream is clearer.  What about `largeStream: Boolean`?

Reply:
yeah, how about "requiresMemoryManagementPerElement"?